### ISSUE TITLE
Understanding AWS IP Insights output score

### ISSUE THREAD

**Issue by ksachdeva11** [awslabs/amazon-sagemaker-examples#1533]
Hello, I was testing out the IP Insights model and had a couple of doubts. It would be really helpful if someone can help me with this.

- After performing a batch transform on the model to predict inference results, the model produces negative as well as positive scores. Which score indicates fraud or malicious activity? Is it the negative score or the positive? Because the model documentation says this "For example when a user attempts to log in from an IP address, if the IP Insights score is high enough, a web login server might decide to trigger a multi-factor authentication system."

- And in the github example (https://github.com/awslabs/amazon-sagemaker-examples/blob/master/introduction_to_amazon_algorithms/ipinsights_login/ipinsights-tutorial.ipynb ), predictions are converted to negative values. What is the reason for that? 

Thanks!

**Comment by ksachdeva11** [comment]
@knakad @zsaziz @SomnathSI @jaredleekatzman - Could you please check this and provide any inputs? Thanks!

**Comment by SomnathSI** [comment]
Thanks for using IPInsights. 
The default output of IP Insights algorithm is the dot_product between the input entity and IP address. The dot_product signifies how compatible the model considers the entity and IP address. Higher the value of dot product, more similar is the input entity and the IP address. Also, the dot_product is unbounded. Please refer the output response format here: https://docs.aws.amazon.com/sagemaker/latest/dg/ip-insights-inference-data-formats.html

To label a malicious traffic we need to find a threshold 'h', such that for score greater than 'h', we assign a label '1' (malicious - no similarity between input entity and the IP address) and 0 otherwise. However since the dot product is a measure of similarity (higher value means more similar) we convert it to a negative values in the notebook. We then use the negative of dot product as a measure of dissimilarity implying a malicious IP. 

To answer your first question, if you set the threshold 'h' as 0, negative values of the dot product will imply malicious IP.  Let me know if that answers both of your questions.

**Comment by ksachdeva11** [comment]
@SomnathSI Thanks for your reply. Below is the prediction output that the model gives to me. 
 ```
{'dot_product': 0.25884348154067993},
  {'dot_product': -0.03124057501554489},
  {'dot_product': -6.488806247711182},
  {'dot_product': -5.719447135925293},
  {'dot_product': -6.7357354164123535},
  {'dot_product': -6.474661350250244},
  {'dot_product': -5.971259593963623}
```

So, from the above output **0.25884348154067993** indicates the least malicious entity-IP pair, and **-6.7357354164123535** indicates the most malicious entity-IP pair, right?

Also, say an IP insights model was trained using **ml.m5.4xlarge** instance type because the training data was huge, do we need the same instance type to perform batch transform on this model, even though the batch transform data is comparatively very less?

I tried running the batch transform using **ml.m4.xlarge** instance type against the model trained on **ml.m5.4xlarge** instance type and it takes a lot of time. I eventually had to kill the job. Is this expected?

Thanks!

**Comment by SomnathSI** [comment]
Yes, to your first question.  

Regarding, the instances for batch transform, IP insights model size can be very big. The model size is determined by two hyperparameters: vector_dim and num_entity_vectors. We provide an instance recommendation matrix based on these hyperparameters here: https://docs.aws.amazon.com/sagemaker/latest/dg/ip-insights.html#ip-insights-instances-cpu

**Comment by ksachdeva11** [comment]
Yes, I used a higher instance type for training because my training data was huge. But, to perform batch transform on the same model, do I still need a higher instance type even if my batch transform data is very less?

**Comment by SomnathSI** [comment]
The instance recommendation matrix is for inference (including batch transform). Note that if your model is large it will consume big part of the instance memory.

**Comment by ksachdeva11** [comment]
Got it! Thanks a lot!
